### PR TITLE
Mutation overview

### LINKED ISSUE #607: Mutation overview in collections
**Aim:** It should be easier to get an overview of the mutations of the variants in a collection.

**Idea:** A new tab "mutations" should be added to the collection pages. It should contain a table with the following columns if no baseline is selected:

- \>90% (which contains fixed mutations/mutations that occur in more than 90% of the sequences)
- 60-90%
- 30-60%
- 5-30%
- "See all"-button that links to the mutations box on the variant page

If a baseline is selected, the table will compare the mutation set of the variants in the collection to the **fixed/>90%** mutations of the baseline. It shows a nested header like this:

image

### PR DESCRIPTION
**Pull request by anastasia-escher**
Resolves #607

### CODE REVIEW

File: src/pages/CollectionSingleViewPage.tsx [hunk 1]
```diff
@@ -359,6 +400,144 @@ const TabPanel = ({ children, value, index, ...other }: TabPanelProps) => {
...
+
+        console.log('baseline', data);
+
+        return data;
+      }
+    },
+    [baselineVariant, mode, locationSelector, dateRangeSelector]
+  );
+
+  // #####################################
```
Comment (fengelniederhammer) [inline src/pages/CollectionSingleViewPage.tsx]: How about extracting some of those hook usages in custom hooks? That would make the function more readable and we don't need this visual separator in the code.

**Comment by fengelniederhammer** [comment]
Maybe add a test for the new components? :)

File: src/data/MutationsOnCollectionPage.ts [hunk 2]
```diff
@@ -0,0 +1,32 @@
+import { MutationProportionEntry } from './MutationProportionEntry';
+import { VariantSelector } from './VariantSelector';
+
```
Comment (chaoran-chen) [inline src/data/MutationsOnCollectionPage.ts]: File names starting with a capital letter usually indicate that it holds a component, type, or class of the same name. That's not the case here.

File: src/components/MutationsTable.tsx [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+      </div>
+      <br />
+      <div style={{ display: 'block', color: 'blue', margin: '10px' }}>
+        <button className='underline break-words overflow-hidden w-full'>
+          <Link
+            to={`/explore/${placeString}/AllSamples/Past6M/variants?${urlParams.toString()}#mutations`}
+            className='overflow-hidden'
+          >
+            See all
+          </Link>
```
Comment (chaoran-chen) [inline src/components/MutationsTable.tsx]: 1. This can be moved into the above link. We don't need a separate link.
2. It should also respect the date filter that we have on the page.
3. When a baseline is selected, it should open the compare page (not "compare to baseline" because the "compare to base line" mode at the moment does not show the mutation differences)

File: src/pages/CollectionSingleViewPage.tsx [hunk 4]
```diff
@@ -359,6 +400,144 @@ const TabPanel = ({ children, value, index, ...other }: TabPanelProps) => {
...
+    }
+  }, [mutations]);
+
+  const baselineTableData = useMemo(() => {
+    if (mode === 'CompareToBaseline' && baselineMutations.data && mutations.data) {
+      const data: BaselineData = {
+        nucMutations: baselineMutations.data.nucMutations,
+        aaMutations: baselineMutations.data.aaMutations,
+      };
+      console.log('baselineMutations', baselineMutations);
```
Comment (chaoran-chen) [inline src/pages/CollectionSingleViewPage.tsx]: Please remove logging

File: src/pages/CollectionSingleViewPage.tsx (hunk 1 above)
Comment (chaoran-chen) [reply src/pages/CollectionSingleViewPage.tsx]: I like the idea!

File: src/pages/CollectionSingleViewPage.tsx [hunk 5]
```diff
@@ -359,6 +400,144 @@ const TabPanel = ({ children, value, index, ...other }: TabPanelProps) => {
...
+        const baselineData = Promise.all([
+          fetchSamplesCount(selector, signal),
+          MutationProportionData.fromApi(selector, 'aa', signal),
+          MutationProportionData.fromApi(selector, 'nuc', signal),
+        ]).then(async ([variantCount, aaMutationDataset, nucMutationDataset]) => {
+          data.aaMutations = aaMutationDataset.payload;
+          data.nucMutations = nucMutationDataset.payload;
+        });
+
+        console.log('baseline', data);
```
Comment (chaoran-chen) [inline src/pages/CollectionSingleViewPage.tsx]: Please remove logging

**Review by chaoran-chen — CHANGES_REQUESTED** [review]
Sorry that it took me so long to review!

Is the compare-to-baseline mode working correctly? From #607: "If a baseline is selected, the table will compare the mutation set of the variants in the collection to the fixed/>90% mutations of the baseline." - Where are the mutations of the baseline filtered?

Comparing [BF.7.14 on this collection page](https://cov-spectrum-develop-git-mutationoverview-cov-spectrum.vercel.app/collections/136) and the [Venn diagrams on the compare variants](https://cov-spectrum-develop-git-mutationoverview-cov-spectrum.vercel.app/collections/136), if the Venn diagram is set to 90%, shouldn't we see the same values in the "Additional >90%" column and in the "BF.7 [...]"-only part of the Venn diagram?

image

image

File: src/components/MutationsTable.tsx [hunk 6]
```diff
@@ -0,0 +1,171 @@
...
+              mutations.map(item => (
+                <tr key={item.name}>
+                  <td>{renderNameCell(item.query, locationSelector, item.name)}</td>
+                  {mutationColumnsProperties.map(field => {
+                    return (
+                      <td>
+                        {mutationType === 'aa'
+                          ? pullAll(
+                              getNames(filterMutations(item.aaMutations, field)),
+                              getNames(filterMutations(baselineData.aaMutations, field))
```
Comment (chaoran-chen) [inline src/components/MutationsTable.tsx]: Isn't this wrong? Isn't this comparing e.g., 60-90% of a variant to 60-90% of the baseline instead of always comparing to >90% as stated in the issue:

> If a baseline is selected, the table will compare the mutation set of the variants in the collection to the fixed/>90% mutations of the baseline.

It would actually be great to add this as an explanation to the table just above the table.

**Review by chaoran-chen — CHANGES_REQUESTED** [review]
In addition to the above comment, I see the following tasks remaining:

- When clicking on the name of a variant:
  - [x] The date filter should be the same
  - [x] If a baseline is selected, it should open the compare page (not "compare to baseline" because the "compare to baseline" mode at the moment does not show the mutation differences)
- [x] Clean up the code and remove the console.logs
- [x] Sort the mutations as usual

@fengelniederhammer @JonasKellerer, you are also welcome to review as well if you like. It's a lot of code and having a few more people looking over it can be helpful.

File: src/components/__tests__/MutationTables.test.tsx [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+        <BaselineMutationsTable
+          baselineData={baselineData}
+          mutations={mutations}
+          mutationType='aa'
+          locationSelector={locationSelector}
+        />
+      </BrowserRouter>
+    );
+
+    const table = getByTestId('baseline-mutation-table');
```
Comment (fengelniederhammer) [inline src/components/__tests__/MutationTables.test.tsx]: IntelliJ complains: "ESLint: Avoid destructuring queries from `render` result, use `screen.getByTestId` instead(testing-library/prefer-screen-queries)"

File: src/components/MutationsTable.tsx [hunk 8]
```diff
@@ -0,0 +1,218 @@
...
+export type BaselineMutationsTableProps = {
+  baselineData: BaselineData;
+  mutations: MutationObject[] | undefined;
+  mutationType: string;
+  locationSelector: LocationSelector;
+  baselineVariant?: VariantSelector;
+  dateRangeSelector: DateRangeSelector;
+};
+
+export const BaselineMutationsTable = ({
```
Comment (fengelniederhammer) [inline src/components/MutationsTable.tsx]: The test complains:
```  
console.error
    Warning: Each child in a list should have a unique "key" prop.
    
    Check the render method of `BaselineMutationsTable`. See https://reactjs.org/link/warning-keys for more information.
```

File: src/components/MutationsTable.tsx [hunk 9]
```diff
@@ -0,0 +1,218 @@
...
+  baselineData,
+  mutations,
+  mutationType,
+  locationSelector,
+  baselineVariant,
+  dateRangeSelector,
+}: BaselineMutationsTableProps) => {
+  const headers: string[] = ['Additional', 'Missing'];
+  console.log('baseline mutations', baselineData);
```
Comment (fengelniederhammer) [inline src/components/MutationsTable.tsx]: Please remove this :)

File: src/data/mutationsOnCollectionPageData.ts [hunk 10]
```diff
@@ -0,0 +1,28 @@
...
+export interface MutationColumnProperty {
+  min: number;
+  max: number;
+  label: string;
+}
+
+export interface MutationObject {
+  query: VariantSelector;
+  name: string;
+  variantCout: number;
```
Comment (fengelniederhammer) [inline src/data/mutationsOnCollectionPageData.ts]: There is a typo: `variantCouNt`

File: src/pages/CollectionSingleViewPage.tsx [hunk 11]
```diff
@@ -359,13 +402,149 @@ const TabPanel = ({ children, value, index, ...other }: TabPanelProps) => {
...
+          variant: baselineVariant,
+          samplingStrategy: SamplingStrategy.AllSamples,
+          dateRange: dateRangeSelector,
+        });
+
+        Promise.all([
+          fetchSamplesCount(selector, signal),
+          MutationProportionData.fromApi(selector, 'aa', signal),
+          MutationProportionData.fromApi(selector, 'nuc', signal),
+        ]).then(async ([variantCount, aaMutationDataset, nucMutationDataset]) => {
```
Comment (fengelniederhammer) [inline src/pages/CollectionSingleViewPage.tsx]: `variantCount` is unused -> Please either use it or delete the unnecessary call.

File: src/pages/CollectionSingleViewPage.tsx [hunk 12]
```diff
@@ -327,6 +343,33 @@ export const CollectionSingleViewPage = () => {
               <Loader />
             )}
           </TabPanel>
+          <TabPanel value={tab} index={2}>
+            <FormControl>
+              <RadioGroup
+                aria-labelledby='demo-radio-buttons-group-label'
+                defaultValue='female'
```
Comment (fengelniederhammer) [inline src/pages/CollectionSingleViewPage.tsx]: This is probably a left over from copy-pasting? :)

File: src/pages/CollectionSingleViewPage.tsx [hunk 13]
```diff
@@ -61,6 +68,7 @@ export const CollectionSingleViewPage = () => {
   const history = useHistory();
   const locationState = useLocation();
   const [tab, setTab] = useState(0);
+  const [mutationType, setMutationType] = useState('aa');
```
Comment (fengelniederhammer) [inline src/pages/CollectionSingleViewPage.tsx]: Can we narrow the type to `useState<'aa' | 'nuc'>`?

File: src/pages/CollectionSingleViewPage.tsx [hunk 14]
```diff
@@ -359,13 +402,149 @@ const TabPanel = ({ children, value, index, ...other }: TabPanelProps) => {
...
+          <div className='mt-4'>
+            <SingleMutationsTable
+              mutations={mutationTableData}
+              locationSelector={locationSelector}
+              mutationType={mutationType}
+              dateRangeSelector={dateRangeSelector}
+            />
+          </div>
+        </>
+      ) : baselineTableData && mode === 'CompareToBaseline' ? (
```
Comment (fengelniederhammer) [inline src/pages/CollectionSingleViewPage.tsx]: Why is it not necessary to check for `mutationTableData` here? I would have expected the check to be `mutationTableData && baselineTableData && mode === 'CompareToBaseline'`

File: src/components/MutationsTable.tsx [hunk 15]
```diff
@@ -0,0 +1,257 @@
...
+
+export const SingleMutationsTable = ({
+  mutations,
+  locationSelector,
+  mutationType,
+  dateRangeSelector,
+}: SingleMutationsTableProps) => {
+  return (
+    <>
+      {mutations && (
```
Comment (fengelniederhammer) [inline src/components/MutationsTable.tsx]: Why do we do the same check twice? The parent component already checks for truthiness of `mutations`

File: src/components/MutationsTable.tsx [hunk 16]
```diff
@@ -0,0 +1,257 @@
...
+import { sortMutationList } from '../helpers/mutation';
+
+type SingleMutationsTableProps = {
+  mutations: MutationObject[] | undefined;
+  locationSelector: LocationSelector;
+  mutationType: string;
+  dateRangeSelector: DateRangeSelector;
+};
+
+const renderNameCell = (
```
Comment (fengelniederhammer) [inline src/components/MutationsTable.tsx]: This is a component. Component names should start with a capital letter. Maybe simply rename this to `NameCell`? Also please use it like a component (props as argument, call it via JSX).

File: src/pages/CollectionSingleViewPage.tsx (hunk 13 above)
Comment (fengelniederhammer) [reply src/pages/CollectionSingleViewPage.tsx]: There is even already the `SequenceType` for that.

File: src/pages/CollectionSingleViewPage.tsx [hunk 17]
```diff
@@ -359,13 +402,149 @@ const TabPanel = ({ children, value, index, ...other }: TabPanelProps) => {
...
+        })
+      );
+
+      return data;
+    },
+    [variants, locationSelector, dateRangeSelector]
+  );
+
+  const baselineMutations = useQuery(
+    async signal => {
```
Comment (fengelniederhammer) [inline src/pages/CollectionSingleViewPage.tsx]: The problem, why sometimes data doesn't load, is: This async function isn't actually async, since it does not `await` anything. 

How to reproduce it consistently:
* choose a collection
* select a baseline
* switch to the "mutations" tab

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/async_function

"Top-level code, up to and including the first await expression (if there is one), is run synchronously. In this way, an async function without an await expression will run synchronously."
-> The function as it is now will initialize an empty data object, construct a promise and immediately return the empty data. 

I'd propose to restructure this function to use the promises in a more idiomatic way:
```
  const baselineMutations = useQuery(
    signal => {
      if (mode !== 'CompareToBaseline') {
        return Promise.resolve({
          aaMutations: [],
          nucMutations: [],
        });
      }

      const selector: LapisSelector = addDefaultHostAndQc({
        location: locationSelector,
        variant: baselineVariant,
        samplingStrategy: SamplingStrategy.AllSamples,
        dateRange: dateRangeSelector,
      });

      return Promise.all([
        MutationProportionData.fromApi(selector, 'aa', signal),
        MutationProportionData.fromApi(selector, 'nuc', signal),
      ]).then(async ([aaMutationDataset, nucMutationDataset]) => {
        return {
          aaMutations: aaMutationDataset.payload,
          nucMutations: nucMutationDataset.payload,
        };
      });
    },
    [baselineVariant, mode, locationSelector, dateRangeSelector]
  );
```

The same applies to the function in the "mutations useQuery". The usage of promises mirrors more closely the flow of the data instead of setting data via side effects. (Also notice that the function doesn't need to be explicitly `async` because it returns a Promise anyway.)

The problem here can also be fixed by simply `await`ing the Promise in the async function.

**Review by fengelniederhammer — CHANGES_REQUESTED** [review]

File: src/components/MutationsTable.tsx [hunk 18]
```diff
@@ -0,0 +1,257 @@
...
+                              )
+                            ).join(', ')}
+                      </td>
+                    );
+                  })}
+
+                  {mutationColumnsProperties.map(field => {
+                    return (
+                      <td key={field.min}>
+                        {mutationType === 'aa'
```
Comment (fengelniederhammer) [inline src/components/MutationsTable.tsx]: There is a lot of duplication in those places. It always follows the same pattern:
`mutationType === 'aa' ? doSomething(item.aaMutations) : doSomething(item.nucMutations)`

So the code does exactly the same (as far as I can judge), except for "either get aa or nuc mutations" and "sort according to aa or nuc".

Could we generalize the code a bit to reduce duplication? Introduce a clcass that wraps around mutationType and the mutation data? But there is no need to do that if it's too much work.

File: src/pages/CollectionSingleViewPage.tsx [hunk 19]
```diff
@@ -359,13 +402,149 @@ const TabPanel = ({ children, value, index, ...other }: TabPanelProps) => {
...
+type MutationTableTabContentProps = {
+  mutationType: string;
+  locationSelector: LocationSelector;
+  variants: { query: VariantSelector; name: string; description: string }[];
+  dateRangeSelector: DateRangeSelector;
+  mode: 'Single' | 'CompareToBaseline';
+  baselineVariant?: VariantSelector;
+};
+
+const MutationTableTabContent = ({
```
Comment (fengelniederhammer) [inline src/pages/CollectionSingleViewPage.tsx]: Can we maybe cut this component in smaller pieces? I'd propose the following structure (in pseudocode):

```
function MutationTableTabContent (props) {
  return mode === 'Single' ? <SingleMutationTableTabContent /> : <BaselineMutationTableTabContent  />
} 

function SingleMutationTableTabContent() {
  data = useMutationsData(selectors, variants, ...)
  return <SingleMutationsTable data={data} />
}

function BaselineMutationTableTabContent() {
  data = useBaselineMutationsData(selectors, variants, ...)
  return <BaselineMutationsTable data={data}
}

function useBaselineMutationsData(...) {
  mutationsData = useMutationsData(selectors, variants, ...)

  // other useQuery, useMemo etc. for baseline data only

  return { mutationsData , baselineData }
}

function useMutationsData(...) {
  // useQuery, useMemo etc.
  return mutationsData 
}
```

Maybe even in a separate file? The current file grew to 800 lines which is quite a lot.

File: src/components/__tests__/MutationTables.test.tsx [hunk 20]
```diff
@@ -0,0 +1,125 @@
...
+      nucMutations: [
+        { proportion: 0.051, mutation: 'C20404T', count: 4 },
+        { proportion: 0.091, mutation: 'T670G', count: 5 },
+      ],
+    },
+  ];
+
+  it('the table is rendered', () => {
+    render(
+      <BrowserRouter>
```
Comment (fengelniederhammer) [inline src/components/__tests__/MutationTables.test.tsx]: Doesn't change much here, but I think it's better to use `MemoryRouter` in tests: https://v5.reactrouter.com/web/guides/testing

File: src/components/VariantMutations.tsx [hunk 21]
```diff
@@ -259,7 +259,7 @@ export const VariantMutations = ({ selector }: Props) => {
 
   return (
     <>
-      <div className='mb-8'>
+      <div className='mb-8' id='mutations'>
```
Comment (fengelniederhammer) [inline src/components/VariantMutations.tsx]: Why do we need this id?

File: src/pages/FocusSinglePage.tsx [hunk 22]
```diff
@@ -511,7 +511,7 @@ export const FocusSinglePageContent = ({
               </Sentry.ErrorBoundary>
             </div>
 
-            <div className='m-4'>
+            <div className='m-4' id='mutations1'>
```
Comment (fengelniederhammer) [inline src/pages/FocusSinglePage.tsx]: Why do we need this id?

File: src/pages/CollectionSingleViewPage.tsx [hunk 23]
```diff
@@ -448,10 +628,13 @@ const TableTabContent = ({
         const query = params.row.query;
         const urlParams = new URLSearchParams();
         addVariantSelectorToUrlSearchParams(query, urlParams);
+
         const placeString = encodeLocationSelectorToSingleString(locationSelector);
         return params.value ? (
           <Link
-            to={`/explore/${placeString}/AllSamples/Past6M/variants?${urlParams.toString()}`}
+            to={`/explore/${placeString}/AllSamples/${dateRangeUrlFromSelector(
+              dateRangeSelector
```
Comment (fengelniederhammer) [inline src/pages/CollectionSingleViewPage.tsx]: Why do we need the dateRangeSelector here now? Was this a bug before?

File: src/pages/CollectionSingleViewPage.tsx (hunk 13 above)
Comment (fengelniederhammer) [inline src/pages/CollectionSingleViewPage.tsx]: If we wrap the `FormControl` and the `MutationTableTabContent` in another component, we can move this state out of this far-too-big component into that other component.

File: src/components/MutationsTable.tsx [hunk 24]
```diff
@@ -0,0 +1,257 @@
...
+  );
+};
+
+const getNames = (mutations: MutationProportionEntry[]) => {
+  return mutations.map(item => item.mutation);
+};
+
+function commonElements<T>(arr1: T[], arr2: T[]): T[] {
+  return arr1.filter(function (element) {
+    return arr2.indexOf(element) !== -1;
```
Comment (fengelniederhammer) [inline src/components/MutationsTable.tsx]: Out of interest: Is `indexOf` better than `includes`? `includes` would make the intention clearer.

Comment (fengelniederhammer) [reply src/components/MutationsTable.tsx]: ~I think one could even shorten this to `return arr1.filter(arr2.includes);` then.~ Somehow doesn't work.

File: src/pages/CollectionSingleViewPage.tsx (hunk 19 above)
Comment (fengelniederhammer) [reply src/pages/CollectionSingleViewPage.tsx]: The downside is that we lose caching that way, because the self-written useQuery doesn't have a cache. However the browser seems to have some basic caching there, so we might save some requests.

Using `react-query` would really help here, because it has a proper built-in cache.

File: src/components/MutationsTable.tsx [hunk 25]
```diff
@@ -0,0 +1,257 @@
...
+                    );
+                  })}
+
+                  {mutationColumnsProperties.map(field => {
+                    return (
+                      <td key={field.min}>
+                        {mutationType === 'aa'
+                          ? sortMutationList(
+                              'aa',
+                              compareToBaseline(
```
Comment (fengelniederhammer) [inline src/components/MutationsTable.tsx]: What is this supposed to do? I don't quite get the meaning of "Missing" yet. Missing from what? What do the percentages mean (90% of what)?

The code selects mutations that are in the baseline and the complement of `field.min` in `item`. How does this connect to the headline "Missing"?

File: src/components/MutationsTable.tsx [hunk 26]
```diff
@@ -0,0 +1,257 @@
...
+                      true,
+                      dateRangeSelector,
+                      baselineVariant
+                    )}
+                  </td>
+                  {mutationColumnsProperties.map(field => {
+                    return (
+                      <td key={field.label}>
+                        {mutationType === 'aa'
+                          ? sortMutationList(
```
Comment (fengelniederhammer) [inline src/components/MutationsTable.tsx]: This component contains both: display of data and the logic. I would prefer to separate both for testability:
* The logic: Given some mutation data, returns data to display -> Test the data selection logic (Given mutation and baseline data, some mutation ends up in the xx% segment of the table.)
* The display: Given the output of the previous function, it renders it as a table. -> Test display logic only (is a nice table).

File: src/components/MutationsTable.tsx (hunk 24 above)
Comment (chaoran-chen) [reply src/components/MutationsTable.tsx]: `arr1.filter(arr2.includes)` looks dangerous because `filter` does not just pass the element to the callback function but actually `(element, index, array)`, and `includes` accepts `fromIndex` as its second argument. I.e., `filter's index` will be used as `includes' fromIndex`

File: src/components/MutationsTable.tsx (hunk 25 above)
Comment (anastasia-escher) [reply src/components/MutationsTable.tsx]: The name of the function was a bit misleading, so I've just renamed it to 'findMissingMutations'. It should find missing mutations based on the following logic (as explained by @chaoran-chen): "If a fixed mutation of the baseline is present in 25% of variant samples - in other words missing in 75% of the samples, it should occur in the column 60-90% but in none of the others."

File: src/pages/CollectionSingleViewPage.tsx (hunk 23 above)
Comment (anastasia-escher) [reply src/pages/CollectionSingleViewPage.tsx]: Yes, the link should open the main page with the date selector specified on the collection page.

File: src/pages/FocusSinglePage.tsx (hunk 22 above)
Comment (anastasia-escher) [reply src/pages/FocusSinglePage.tsx:514]: The link to the main page from the Mutation Tab should open the main page on the section with mutations, so this id was added in order to use Hashlink.

File: src/components/VariantMutations.tsx (hunk 21 above)
Comment (anastasia-escher) [reply src/components/VariantMutations.tsx]: The link form the mutations tab should lead to the mutations section on the main page, so we need it in order to use a hash link.

Comment (fengelniederhammer) [reply src/components/VariantMutations.tsx]: It works, but only if the page is already fully loaded. When the page is still building up, I end up in the wrong place on the page.

File: src/components/MutationsTable.tsx (hunk 15 above)
Comment (anastasia-escher) [reply src/components/MutationsTable.tsx]: Without this check I get the "the object is probably undefined" error in mutations.map(...).

Comment (fengelniederhammer) [reply src/components/MutationsTable.tsx]: You can require in the Props type that it is a `MutationObject[]` instead of `MutationObject[] | undefined`

File: src/components/MutationTableTabContent.tsx [hunk 27]
```diff
@@ -66,31 +66,30 @@ export default function MutationTableTabContent({
...
+        samplingStrategy: SamplingStrategy.AllSamples,
+        dateRange: dateRangeSelector,
+      });
 
-        return data;
-      }
+      return Promise.all([
+        MutationProportionData.fromApi(selector, 'aa', signal),
+        MutationProportionData.fromApi(selector, 'nuc', signal),
+      ]).then(async ([aaMutationDataset, nucMutationDataset]) => {
```
Comment (fengelniederhammer) [inline src/components/MutationTableTabContent.tsx]: This doesn't need to be async.

File: src/pages/CollectionSingleViewPage.tsx [hunk 28]
```diff
@@ -325,7 +325,6 @@ export const CollectionSingleViewPage = () => {
             <FormControl>
               <RadioGroup
                 aria-labelledby='demo-radio-buttons-group-label'
-                defaultValue='female'
                 name='radio-buttons-group'
```
Comment (fengelniederhammer) [inline src/pages/CollectionSingleViewPage.tsx]: Do we need the `aria-labelledby` and the `name`? Or could we delete those, too?

File: src/components/MutationTableTabContent.tsx [hunk 29]
```diff
@@ -0,0 +1,146 @@
...
+          name: variant.name.length > 0 ? variant.name : formatVariantDisplayName(variant.query),
+          variantCount: variant.variantCount,
+          aaMutations: variant.aaMutations,
+          nucMutations: variant.nucMutations,
+        };
+      });
+    }
+  }, [mutations]);
+
+  const baselineTableData = useMemo(() => {
```
Comment (fengelniederhammer) [inline src/components/MutationTableTabContent.tsx]: Why do we actually need a `useMemo` here? Couldn't we simply write 
```
const baselineTableData = mode === 'CompareToBaseline' && baselineMutations.data && mutations.data 
  ? baselineMutations.data 
  : undefined
``` 
?

Comment (fengelniederhammer) [reply src/components/MutationTableTabContent.tsx]: Why do we need that check at all? Isn't is essentially the same check as line 132 below?

File: src/components/MutationsTable.tsx [hunk 30]
```diff
@@ -0,0 +1,271 @@
...
+  baseline,
+  dateRangeSelector,
+  baselineVariant,
+}: NameCellProps) => {
+  const urlParams = new URLSearchParams();
+  addVariantSelectorToUrlSearchParams(query, urlParams);
+
+  const placeString = encodeLocationSelectorToSingleString(locationSelector);
+  return (
+    <>
```
Comment (fengelniederhammer) [inline src/components/MutationsTable.tsx]: We don't need that fragment

File: src/components/MutationsTable.tsx [hunk 31]
```diff
@@ -0,0 +1,271 @@
...
+const columnNames: string[] = mutationColumnsProperties.map(item => item.label);
+
+export const SingleMutationsTable = ({
+  mutations,
+  locationSelector,
+  mutationType,
+  dateRangeSelector,
+}: SingleMutationsTableProps) => {
+  return (
+    <>
```
Comment (fengelniederhammer) [inline src/components/MutationsTable.tsx]: We don't need that fragment

File: src/components/MutationsTable.tsx [hunk 32]
```diff
@@ -0,0 +1,271 @@
...
+export const BaselineMutationsTable = ({
+  baselineData,
+  mutations,
+  mutationType,
+  locationSelector,
+  baselineVariant,
+  dateRangeSelector,
+}: BaselineMutationsTableProps) => {
+  const headers: string[] = ['Additional', 'Missing'];
+  console.log('baseline data', baselineData);
```
Comment (fengelniederhammer) [inline src/components/MutationsTable.tsx]: Please delete that log statement before merge.

File: src/components/MutationsTable.tsx [hunk 33]
```diff
@@ -0,0 +1,271 @@
...
+              {columnNames.map((name, i) => (
+                <td key={`${name}-${i}`}>{name}</td>
+              ))}
+              {columnNames.map((name, i) => (
+                <td key={`${name}/${i}`}>{name}</td>
+              ))}
+            </tr>
+          </thead>
+          <tbody>
+            {mutations &&
```
Comment (fengelniederhammer) [inline src/components/MutationsTable.tsx]: Same here: Please require in the Props that mutations is not undefined and remove this check here.

File: src/components/VariantMutations.tsx (hunk 21 above)
Comment (fengelniederhammer) [reply src/components/VariantMutations.tsx]: I don't know how to fix this or whether it is possible to fix this at all with simple means. I guess that the issue is that the page loads tons of data and builds up incrementally, thus it scroll to the "correct" place, but after loading the next diagram, the actually correct place is further down.
How much of a hard requirement is this? Could we simply leave out this feature? Or implement something else instead (only showing the targetted component)? I think we should at least not leave a non-working feature here.

File: src/components/MutationsTable.tsx (hunk 25 above)
Comment (fengelniederhammer) [reply src/components/MutationsTable.tsx]: Ok, makes sense. Please add a description to the table to clarify this. @chaoran-chen said he will provide a text at the beginning of next week.

File: src/data/api-lapis.ts [hunk 34]
```diff
@@ -27,15 +27,12 @@ import { LapisSelector } from './LapisSelector';
...
 import { addQcSelectorToUrlSearchParams } from './QcSelector';
 import { HostCountSampleEntry } from './sample/HostCountSampleEntry';
-import { sequenceDataSource } from '../helpers/sequence-data-source';
 import { InsertionCountEntry } from './InsertionCountEntry';
 import { NextcladeDatasetInfo } from './NextcladeDatasetInfo';
 
 const HOST = process.env.REACT_APP_LAPIS_HOST;
 const ACCESS_KEY = process.env.REACT_APP_LAPIS_ACCESS_KEY;
 
-export const HUMAN = sequenceDataSource === 'gisaid' ? 'Human' : 'Homo sapiens';
```
Comment (fengelniederhammer) [inline src/data/api-lapis.ts:37]: I moved this to resolve a circular module dependency

File: src/components/TableTabContent.tsx [hunk 35]
```diff
@@ -0,0 +1,263 @@
+import { DataGrid, GridColDef, GridComparatorFn, GridRenderCellParams } from '@mui/x-data-grid';
```
Comment (fengelniederhammer) [inline src/components/TableTabContent.tsx:1]: This was extracted from the `CollectionSingleViewPage`

File: src/components/SequencesOverTimeTabContent.tsx [hunk 36]
```diff
@@ -0,0 +1,103 @@
+import { useRef } from 'react';
```
Comment (fengelniederhammer) [inline src/components/SequencesOverTimeTabContent.tsx:1]: This was extracted from the `CollectionSingleViewPage`

File: src/components/VariantMutations.tsx (hunk 21 above)
Comment (fengelniederhammer) [reply src/components/VariantMutations.tsx]: I set a timeout after which the scroll effect happens.

**Comment by fengelniederhammer** [comment]
@chaoran-chen Could you please provide the explanation text for the baseline comparison table? Otherwise this PR is ready from our side and waits for a review :)

Sorry that this PR got that messy, I had to rebase it and squashed all commits. What this PR includes at this point:
* some test setup: Mock the API calls that are issued at static "load the file" time. This should be refactored at some point in the future.
* resolving circular dependencies of modules that somehow broke the page (I don't know why it worked before - there are still some circular dependencies)
* The new mutation tables - structured in hooks that compute the data and components that only display a table, including tests.
* Extraction of some components to make the `CollectionSingleViewPage` smaller.

**Review by fengelniederhammer — APPROVED** [review]

**Comment by chaoran-chen** [comment]
@fengelniederhammer, I added the text to the mutation table.

**Review by chaoran-chen — APPROVED** [review]
